torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
-        "device_print",
-        "device_assert",
-    }
+class Op:
+    name: str
+    args: List[Union[Param, Intermediate]]
 
-    def __init__(self, infos) -> None:
-        super().__init__()
-        self.infos = infos
-        self.read_depth = 0
-        self.in_store = False
-
-    def visit_Name(self, node):
-        if node.id not in self.infos:
-            return
-        if self.read_depth:
-            pass
-        elif self.in_store:
-            self.infos[node.id].mutated = True
-        else:
-            self.infos[node.id].used_in_unknown = True
-
-    def visit_Call(self, node):
-        # TODO(oulgen): Here we assume that there exists a line called
-        # from triton import language as tl. This needs to be checked
-        # as if someones imports xyz as tl then we will incorrectly
-        # assume a mutation but this would be ok as it is only unsafe to
-        # miss a mutation.
-        if (
-            isinstance(node.func, ast.Attribute)
-            and isinstance(node.func.value, ast.Name)
-            and node.func.value.id == "tl"
+
+# Given a Triton emitted TTIR text, this function lexes and parses the
+# code using a minimal grammar defined inside. During the lexing/parsing,
+# we drop any constant value and type information as they are not
+# necessary to us.
+# Being able to choose what we need makes this not a general purpose TTIR
+# parser which further makes parsing much simpler.
+
+
+def parse_ttir(ttir, kwargs):
+    # TODO(oulgen):
+    # - Support multiple functions
+    # - Support parsing of conditionals
+    # - Support parsing for/while loops
+    # - Support alternative syntax parsing similar to "tt.name"(%0, %1)
+
+    if ttir.count("tt.func") != 1:
+        log.debug("Multiple functions in TTIR")
+        return None
+
+    try:
+        import lark
+        from lark import Lark, Transformer, v_args
+    except ModuleNotFoundError:
+        log.debug("Cannot parse TTIR as Lark is not installed")
+
+    ops: Dict[Intermediate, Op] = {}
+    next_fake_intermediate = 0
+
+    # Ops looks like one of the following forms:
+    #
+    # %14 = tt.addptr %13, %4 : tensor<4x!tt.ptr<f32, 1>>, tensor<4xi32>
+    # tt.store %14, %12, %5 {cache = 1 : i32, evict = 1 : i32} : tensor<4xf32>
+    # %15 = "tt.atomic_rmw"(%14, %12, %5) <{atomic_rmw_op = 5 : i32, scope = 1 : i32, sem = 4 : i32}> : (tensor<4x!tt.ptr<f32, 1>>, tensor<4xf32>, tensor<4xi1>) -> tensor<4xf32>  # noqa: B950
+    grammar = """
+        start: module_block
+
+        module_block: "module" "{" decl_block "}"
+
+        decl_block: /.+/ NEWLINE op+ "}"
+
+        op: "tt.return"
+          | assign_lhs "=" FN_NAME args /.+/ NEWLINE  -> process_op
+          | FN_NAME args /.+/ NEWLINE  -> process_op_no_ret
+
+        args: arg?
+            | arg "," args
+
+        ?arg: INTERMEDIATE | CONSTANT | PARAM
+
+        ?assign_lhs: INTERMEDIATE | CONSTANT
+
+        PARAM.5: "%arg" DIGIT+
+
+        INTERMEDIATE: "%" DIGIT+
```

Comment:
I think you're missing multiple return values, e.g.
```
    %4:2 = "tt.reduce"(%3#0, %3#1) <{axis = 1 : i32}> ({
```

`%4:2` means there are two values returned, which are referenced as `%4#0` and `%4#1`.

This shows up in reductions, or more complicated structured control flow situations.

Reply:
I will add this as a missing feature above. I also did not implement parsing for control flow yet. Just wanted to see what everyone thinks. 

The fact that TTIR can randomly emit both of the syntaxes in the same code is pretty annoying.

```
%14 = tt.addptr %13, %4 : tensor<4x!tt.ptr<f32, 1>>, tensor<4xi32>
%15 = "tt.atomic_rmw"(%14, %12, %5) <{atomic_rmw_op = 5 ...
```